### PR TITLE
✨ Is922/add autorecharge functionality

### PR DESCRIPTION
**Pull request by matusdrobuliak66**
## What do these changes do?
- ✨ Introduce auto recharge functionality to the **Payments service**

*NOTE*: Introduced variable `PAYMENTS_AUTORECHARGE_ENABLED` to the **Payments service**. Based on this variable we pay via the payment gateway (*pay_with_payment_method*). It is a temporary solution so we can turn on/off this functionality.

- <ins>Auto recharge concept</ins>: **Payment service** subscribes to RabbitMQ and listens on the wallet credits events. When the message comes, the **Payment service** starts to process the message, there are a series of steps to check whether the auto recharge should be done, if some of them are not met the auto recharge is not done.
  - Step 1: Check if wallet credits are below the threshold
  - Step 2: Check auto-recharge conditions (whether auto-recharge is enabled for the concrete wallet)
  - Step 3: Get Payment method
  - Step 4: Check spending limits (if the user set monthly limits)
  - Step 5: Check the last top-up time (Protective measure: As **Payment service** can in theory receive multiple messages, we need to make sure, that we do not initialize top-up multiple times. The current approach is that we allow just one top-up in 5 minutes, which should be reasonable for now)
  - Step 6: Perform auto-recharge

## Related issue/s
- relates to https://github.com/ITISFoundation/osparc-issues/issues/1027
- relates to https://github.com/ITISFoundation/osparc-issues/issues/922

## How to test
**Webserver:**
`pytest tests/unit/test_services_auto_recharge_listener.p`

**Payments service:**
`pytest tests/unit/with_dbs/03/products/test_products_rpc.py`

## Dev Checklist

- [x] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))
https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/merge_requests/158 @YuryHrytsuk @mrnicegyu11 

## DevOps Checklist
Added to payments service:
- `PAYMENTS_AUTORECHARGE_DEFAULT_MONTHLY_LIMIT`=10000
- `PAYMENTS_AUTORECHARGE_DEFAULT_TOP_UP_AMOUNT`=100.0
- `PAYMENTS_AUTORECHARGE_MIN_BALANCE_IN_CREDITS`=100
- `PAYMENTS_AUTORECHARGE_ENABLED`=1

### CODE REVIEW

File: packages/models-library/src/models_library/products.py
Comment (pcrespov) [inline packages/models-library/src/models_library/products.py:10]: Annotate more precisely

```python
product_name :ProductName
credit_amount: AmountDecimal
```

File: packages/models-library/src/models_library/rabbitmq_messages.py
Comment (pcrespov) [inline packages/models-library/src/models_library/rabbitmq_messages.py:267]: `product_name: ProductName`

File: services/payments/src/simcore_service_payments/services/payments_methods.py
Comment (pcrespov) [inline services/payments/src/simcore_service_payments/services/payments_methods.py:151]: hehe, we go on and off with the names. I had these names initially and change them because at this interface I wanted to remove the concept of "state". 

A payment-method in this interface "exists" when it is fully created. the state of creation is something delegated to the repo interface but not here. 

With payments is different because you accept, cancel or reject a payment while with payment-methods we only intend to store them.

**Review by pcrespov — COMMENTED** [review]
Looks really good. some initial comments. I will follow later ...

File: services/payments/tests/unit/conftest.py
Comment (sanderegg) [inline services/payments/tests/unit/conftest.py:417]: nice

File: services/payments/tests/unit/test_services_auto_recharge_listener.py
Comment (sanderegg) [inline services/payments/tests/unit/test_services_auto_recharge_listener.py:170]: consider separating in 2 fixtures. this one contains the server but you are returning the client

File: services/payments/tests/unit/test_services_auto_recharge_listener.py
Comment (sanderegg) [inline services/payments/tests/unit/test_services_auto_recharge_listener.py:220]: for example you have it here, you could consider checking whether it was called but since you return the client, you can't

**Review by sanderegg — APPROVED** [review]
very nice! thanks

File: packages/pytest-simcore/src/pytest_simcore/helpers/rawdata_fakers.py
Comment (GitHK) [inline packages/pytest-simcore/src/pytest_simcore/helpers/rawdata_fakers.py:202]: don't we have an enum for this?

File: packages/pytest-simcore/src/pytest_simcore/helpers/rawdata_fakers.py
Comment (GitHK) [inline packages/pytest-simcore/src/pytest_simcore/helpers/rawdata_fakers.py:231]: same here

File: services/payments/src/simcore_service_payments/services/payments_methods.py
Comment (GitHK) [reply services/payments/src/simcore_service_payments/services/payments_methods.py:151]: I'm also confused by this successful in the name. I don't get it how a payment method can be successful.

File: services/payments/tests/unit/conftest.py
Comment (GitHK) [inline services/payments/tests/unit/conftest.py:407]: I would go for the full name like we do with all other services
```suggestion
def mock_resoruce_usage_tracker_service_api(
```

File: services/payments/tests/unit/test_services_auto_recharge_listener.py
Comment (GitHK) [reply services/payments/tests/unit/test_services_auto_recharge_listener.py:170]: @sanderegg guess who influenced this one 🙃

**Review by GitHK — APPROVED** [review]

File: services/payments/src/simcore_service_payments/services/payments_methods.py
Comment (matusdrobuliak66) [reply services/payments/src/simcore_service_payments/services/payments_methods.py:151]: Ok, I have reverted the naming back :) The first time I was working with it, I thought it would be better as I was surprised that in the Database layer, we filter this state and I do not know about it from the higher interface. But I guess when we want to remove the concept of "state" than it doesn't matter.

File: services/payments/tests/unit/conftest.py
Comment (matusdrobuliak66) [reply services/payments/tests/unit/conftest.py:417]: This was just a rearrangement of code that was already there :) but I agree

File: services/payments/tests/unit/test_services_auto_recharge_listener.py
Comment (matusdrobuliak66) [reply services/payments/tests/unit/test_services_auto_recharge_listener.py:170]: Done :+1: and yes kudos to @GitHK :)

File: packages/pytest-simcore/src/pytest_simcore/helpers/rawdata_fakers.py
Comment (matusdrobuliak66) [reply packages/pytest-simcore/src/pytest_simcore/helpers/rawdata_fakers.py:202]: yes we have, updated :+1:

File: packages/pytest-simcore/src/pytest_simcore/helpers/rawdata_fakers.py
Comment (matusdrobuliak66) [reply packages/pytest-simcore/src/pytest_simcore/helpers/rawdata_fakers.py:231]: yes we have, updated :+1:

**Comment by matusdrobuliak66** [comment]
@pcrespov I will wait for your approval

File: services/docker-compose.yml
Comment (YuryHrytsuk) [inline services/docker-compose.yml:276]: Thank you for making it explicit. Let's follow this way :+1:

**Review by YuryHrytsuk — APPROVED** [review]
Gut gemacht. Thank you!

**Review by pcrespov — APPROVED** [review]
Fantastic! :tada:

**Pull request merged**
